### PR TITLE
chore: Add a test workflow

### LINKED ISSUE #680: Migrate GitLab CI tests to GitHub Actions
https://github.com/meltano/sdk/blob/050f604a190866fb16cc2fd1c5697d56010f0216/.gitlab-ci.yml#L55-L81

### PR DESCRIPTION
**Pull request by edgarrmondragon**
This is the first PR in a series to migrate CI jobs from GitLab CI to GitHub Actions.

Checks run:

- Unit tests (`tests/core`, `tests/cookiecutters`) on all supported Python versions
- External tests (`tests/external`). Requires setting up secrets.
- Doctests
- Lint

For unit and external tests, coverage is computed and later combined and uploaded to Codecov in the `coverage` job.

This PR also adds dependabot updates for Python tools used in CI (`poetry`, `pip`) and third-party GitHub actions.

Closes #680

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    - name: Upgrade pip
+      env:
+        PIP_CONSTRAINT: .github/workflows/constraints.txt
+      run: |
+        pip install pip
+        pip --version
+
+    - name: Install Dependencies
+      run: |
+        poetry env use "${{ matrix.python-version }}"
```
Comment (edgarrmondragon) [inline .github/workflows/test.yml:51]: As suggested in https://github.com/actions/setup-python/issues/374#issuecomment-1088938718

File: tox.ini [hunk 2]
```diff
@@ -33,6 +33,12 @@ commands =
     poetry run coverage html
     open ./htmlcov/index.html
 
+[testenv:no_external]
+commands =
+    poetry run pytest --doctest-modules singer_sdk
+    poetry run pytest tests/core
+    poetry run pytest tests/cookiecutters
+
```
Comment (edgarrmondragon) [inline tox.ini:41]: We should probably use [pytest markers](https://docs.pytest.org/en/stable/example/markers.html#marking-individual-tests-when-using-parametrize) to (de)select external tests.

**Review by aaronsteers — APPROVED** [review]
LGTM! Thanks for this quick pass!

One failing CI test and then this is ready to go. Thanks!

I love the codecov display - thanks for that!

**Comment by edgarrmondragon** [comment]
> Do I need to take some action for that?

@aaronsteers yup, any secrets used in the GitLab CI job should be copied to https://github.com/meltano/sdk/settings/secrets/actions

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    continue-on-error: ${{ matrix.external || false }}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+        - { tox-env: py, python-version: "3.7", os: "ubuntu-latest" }
+        - { tox-env: py, python-version: "3.8", os: "ubuntu-latest" }
+        - { tox-env: py, python-version: "3.9", os: "ubuntu-latest" }
+        - { tox-env: py, python-version: "3.10", os: "ubuntu-latest" }
+        - { tox-env: py, python-version: "3.10", os: "ubuntu-latest", external: true }
```
Comment (aaronsteers) [inline .github/workflows/test.yml:23]:
```suggestion
        #TODO: resolve by mapping CI vars to creds: 
        # https://github.com/meltano/sdk/issues/683
        # - { tox-env: py, python-version: "3.10", os: "ubuntu-latest", external: true }
```

**Comment by aaronsteers** [comment]
> > Do I need to take some action for that?
> 
> @aaronsteers yup, any secrets used in the GitLab CI job should be copied to https://github.com/meltano/sdk/settings/secrets/actions

Secrets are now created as organization secrets, shared with this repo and present at the link above. 

I've temporarily removed the `external` item from the matrix so this can pass.

@edgarrmondragon - can we merge for now and handle the follow-up in #683 ?

**Comment by edgarrmondragon** [comment]
@aaronsteers 

> can we merge for now and handle the follow-up in https://github.com/meltano/sdk/issues/683 ?

Sure 👍

**Pull request merged**